application-operator/mcagent/mcagent_syncer.go

```
@@ -177,10 +195,52 @@ func (s *Syncer) updateVMCStatus() error {
 	// If Thanos is disabled, we want to empty the host so Prometheus federation returns
 	vmc.Status.ThanosQueryStore = thanosAPIHost
 
+	// Update the Kubernetes version on the VMC
+	k8sVersion, err := s.getWorkloadK8sVersion()
+	if err != nil {
+		return fmt.Errorf("Failed to get Kubernetes information to update VMC %s: %v", vmcName, err)
+	}
+	vmc.Status.Kubernetes.Version = k8sVersion
+
+	// Update the Verrazzano version on the VMC
+	vzVersion, err := s.getWorkloadVZVersion()
+	if err != nil {
+		return fmt.Errorf("Failed to get Verrazzano information to update VMC %s: %v", vmcName, err)
+	}
+	vmc.Status.Verrazzano.Version = vzVersion
+
 	// update status of VMC
 	return s.AdminClient.Status().Update(s.Context, &vmc)
 }
 
+// getWorkloadK8sVersion retrieves the current Kubernetes version on this managed cluster
+func (s *Syncer) getWorkloadK8sVersion() (string, error) {
+	discoveryClient, err := getDiscoveryClientFunc()
+	if err != nil {
+		return "", fmt.Errorf("failed to get discovery client for this workload cluster: %v", err)
+	}
+
+	k8sVersion, err := discoveryClient.ServerVersion()
+	if err != nil {
+		return "", fmt.Errorf("failed to get Kubernetes version on this workload cluster: %v", err)
+	}
+	return k8sVersion.String(), nil
+}
+
+// getWorkloadVZVersion retrieves the current VZ version on this managed cluster,
+// as reported by the VZ CR's status
+func (s *Syncer) getWorkloadVZVersion() (string, error) {
+	vzList := &v1beta1.VerrazzanoList{}
+	if err := s.LocalClient.List(s.Context, vzList, &client.ListOptions{}); err != nil {
+		return "", fmt.Errorf("error listing Verrazzanos: %v", err)
+	}
+	if len(vzList.Items) != 1 {
```

Is it ever valid for there to be 0 verrazzanos on the managed cluster?

---

it seems like if there were, for some reason, the verrazzano version would be nil

---

It should be possible for there to be 0 verrazzanos on a managed cluster in general. However, I don't think this agent code would run in that case, unless it is possible for the verrazzano-cluster-agent to run without a Verrazzano installation on that cluster.

---

At this point, it is (nearly) impossible for that to be the case.

---

What does "nearly" mean? I suppose that means I should probably at least take the 0 verrazzanos case into account here?

---

My question was somewhat rhetoric, sorry. I mean in general, is it better to return an error here, or nil/empty string if this happens? What would be more defensible

---

I would want to return nil in the 0 Verrazzanos case.
The reasoning is that returning an error would prevent the VMC on the cluster from actually being updated with all the previous changes. See this line of code: https://github.com/verrazzano/verrazzano/blob/mmgellec/VZ-11008/application-operator/mcagent/mcagent_syncer.go#L213.

---

Perhaps since the 0 Verrazzano case is practically impossible, I can plan to include this change in a follow-up, along with the changes about the `Syncer` initialization that were mentioned above.

---

for sure, i'm just being nit-picky

---

I appreciate it. Nit-picky is better than not reviewing the code well